liblwgeom/measures3d.c

```diff
@@ -361,29 +362,66 @@ lwgeom_solid_contains_lwgeom(const LWGEOM *solid, const LWGEOM *g)
 	else /* Raycast upwards in Z */
 	{
 		POINT4D pt;
-		LWCOLLECTION *c;
-		/* take first point */
-		if (!lwgeom_startpoint(g, &pt))
+		/* We'll skew copies if we're not lucky */
+		LWGEOM * solid_copy = lwgeom_clone_deep(solid);
+		LWGEOM * g_copy = lwgeom_clone_deep(g);
+
+		while (LW_TRUE)
 		{
-			return LW_FALSE;
-		};
-		uint8_t is_inside = LW_FALSE;
+			uint8_t is_boundary = LW_FALSE;
+			uint8_t is_inside = LW_FALSE;
 
-		/* get part of solid that is upwards */
-		c = lwgeom_clip_to_ordinate_range(solid, 'Z', pt.z, DBL_MAX, 0);
+			/* take first point */
+			if (!lwgeom_startpoint(g_copy, &pt))
+				return LW_FALSE;
 
-		for (uint32_t i = 0; i < c->ngeoms; i++)
-		{
-			/* walls of solid should be polygons, otherwise uninteresting */
-			if (lwgeom_get_type(c->geoms[i]) != POLYGONTYPE)
-				continue;
+			/* get part of solid that is upwards */
+			LWCOLLECTION *c = lwgeom_clip_to_ordinate_range(solid_copy, 'Z', pt.z, DBL_MAX, 0);
+
+			for (uint32_t i = 0; i < c->ngeoms; i++)
+			{
+				/* walls of solid should be polygons, otherwise uninteresting */
+				if (c->geoms[i]->type != POLYGONTYPE)
+					continue;
+
+				/* 3D raycast along Z is 2D point in polygon */
+				int t = lwpoly_contains_point((LWPOLY *)c->geoms[i], (POINT2D *)&pt);
+
+				if (t == LW_INSIDE)
+					is_inside = !is_inside;
+				else if (t == LW_BOUNDARY)
+				{
+					is_boundary = LW_TRUE;
+					break;
+				}
+			}
+
+			lwcollection_free(c);
+
+			if (is_boundary)
+			{
+				/* randomly skew a bit and restart*/
+				double cx = lwrandom_uniform()-0.5;
```

Comment:
To make this random, `lwrandom_set_seed(0)` should be done somewhere at the top of the `while` loop.

Reply:
At the same time, `ST_3DIntersects` can't be IMMUTABLE if it relies on random behaviour. Could always setting the same seed be the way [to go](https://xkcd.com/221/) as it doesn't need true randomness, just a series of pseudo-random numbers?

Reply:
The randomness is not supposed to change the result. It will change the execution time, not the result. So immutable could stay I think.
However, I agree we do not need true randomness here.